### ISSUE TITLE
[BUG] Brightness down not working on Ikea E1810 & E1743 ZHA

### ISSUE THREAD

**Issue by raghavrithvik** [xaviml/controllerx#457]
# Bug report

## Description

Everything but the the brightness down button on both E1810 & E1743 (4 units) don't seem to work. Even with merge mapping. 

I've confirmed that the zha_event is received on HA, But the AppDaemon logs don't show anything for that button press.

## Additional information

* Devices involved:
  * Model: E1810 & E1743
  * ...
* Integration: Choose from `zha`
* AppDaemon version: 4.2.1
* ControllerX version: 4.18.2
* HACS version (if installed from there): 1.24.3
* Home Assistant Core version: 2022.4.1

## AppDaemon app configuration

```yaml
foyer_ikea_dimmer_2:
  module: controllerx
  class: E1743Controller
  controller: 
  integration: zha
  light: light.foyer_down_lights

livingroom_ikea_5button:
  module: controllerx
  class: E1810Controller
  controller: 
  integration: zha
  light: light.living_room_down_lights
```

## Logs

```text


2022-04-12 16:12:16.758039 INFO livingroom_ikea_5button: 🏃 Running `Predefined (release)` now
--
2022-04-12 16:12:16.749551 INFO livingroom_ikea_5button: 🎮 Button event triggered: `stop`
2022-04-12 16:12:15.008347 INFO livingroom_ikea_5button: 🏃 Running `Predefined (release)` now
2022-04-12 16:12:14.999302 INFO livingroom_ikea_5button: 🎮 Button event triggered: `stop`
2022-04-12 16:12:12.429111 INFO livingroom_ikea_5button: 🏃 Running `Predefined (release)` now
2022-04-12 16:12:12.419935 INFO livingroom_ikea_5button: 🎮 Button event triggered: `stop`
2022-04-12 16:12:10.063373 INFO livingroom_ikea_5button: 🤖 Service: [1mlight.turn_on[0m - entity_id: light.living_room_down_lights - color_temp: 352.30 - transition: 0.30
2022-04-12 16:12:10.039859 INFO livingroom_ikea_5button: 🏃 Running `Predefined (click_color_down)` now
2022-04-12 16:12:10.031101 INFO livingroom_ikea_5button: 🎮 Button event triggered: `press_257_13_0`
2022-04-12 16:12:08.740953 INFO livingroom_ikea_5button: 🤖 Service: [1mlight.turn_on[0m - entity_id: light.living_room_down_lights - color_temp: 387.70 - transition: 0.30
2022-04-12 16:12:08.713720 INFO livingroom_ikea_5button: 🏃 Running `Predefined (click_color_up)` now
2022-04-12 16:12:08.706206 INFO livingroom_ikea_5button: 🎮 Button event triggered: `press_256_13_0`
2022-04-12 16:12:06.533361 INFO livingroom_ikea_5button: 🤖 Service: [1mlight.turn_on[0m - entity_id: light.living_room_down_lights - brightness: 255 - transition: 0.30
2022-04-12 16:12:06.506307 INFO livingroom_ikea_5button: 🏃 Running `Predefined (click_brightness_up)` now
2022-04-12 16:12:06.498528 INFO livingroom_ikea_5button: 🎮 Button event triggered: `step_with_on_off_0_43_5`


```

## Additional Context

Add any other context or screenshots about the bug here.

**Comment by mjmcg** [comment]
Volume down is also not working on the E1810MediaPlayerController. It appears the zha_event args don't match the mapping. The mapping is ` "step_1_43_5": MediaPlayer.CLICK_VOLUME_DOWN`, but the event log below shows two extra arguments. 

As a fix, I tried updating my local ikea.py file to ` "step_1_43_5_0_0": MediaPlayer.CLICK_VOLUME_DOWN`, but no luck. I'm on Home Assistant 2022.4.2.

Any suggestions? Thank for any help!

`{
    "event_type": "zha_event",
    "data": {
        "device_ieee": "2c:11:65:ff:fe:c5:f1:88",
        "unique_id": "2c:11:65:ff:fe:c5:f1:88:1:0x0008",
        "device_id": "809631773d003a153d1e35090fc45cc8",
        "endpoint_id": 1,
        "cluster_id": 8,
        "command": "step",
        "args": [
            1,
            43,
            5,
            0,
            0
        ],
        "params": {
            "step_mode": 1,
            "step_size": 43,
            "transition_time": 5,
            "options_mask": 0,
            "options_override": 0
        }
    },
    "origin": "LOCAL",
    "time_fired": "2022-04-12T23:18:50.587170+00:00",
    "context": {
        "id": "4bb7699cdc5203b53b081e12de62bf6d",
        "parent_id": null,
        "user_id": null
    }
}`

**Comment by MagicMicky** [comment]
Same issue with a `E1744`, the stop event seem to be passed to ControllerX but not the move events

```yaml
desk_music_controller_bedroom:
  module: controllerx
  class: E1744MediaPlayerController
  controller: "redacted"
  integration: zha
  media_player: media_player.bedroom
  actions:
    - move_1_195
    - move_0_195
    - stop
```


Logs from ControllerX / AppDaemon
```
2022-04-13 10:44:22.168667 INFO desk_music_controller_bedroom: 🏃 Running `Predefined (release)` now
2022-04-13 10:44:22.166580 INFO desk_music_controller_bedroom: 🎮 Button event triggered: `stop`
2022-04-13 10:44:15.571019 INFO desk_music_controller_bedroom: 🏃 Running `Predefined (release)` now
2022-04-13 10:44:15.568840 INFO desk_music_controller_bedroom: 🎮 Button event triggered: `stop`
```

Logs when listening to events in HA (zha_event)
Rotation in a direction
```json
{
    "event_type": "zha_event",
    "data": {
        "device_ieee": "redacted",
        "unique_id": "redacted",
        "device_id": "redacted",
        "endpoint_id": 1,
        "cluster_id": 8,
        "command": "move",
        "args": [
            1,
            195,
            0,
            0
        ],
        "params": {
            "move_mode": 1,
            "rate": 195,
            "options_mask": 0,
            "options_override": 0
        }
    },
    "origin": "LOCAL",
    "time_fired": "2022-04-13T08:44:21.976066+00:00",
    "context": {
        "id": "redacted",
        "parent_id": null,
        "user_id": null
    }
}
```
Stop event:
```json
{
    "event_type": "zha_event",
    "data": {
        "device_ieee": "redacted",
        "unique_id": "redacted",
        "device_id": "redacted",
        "endpoint_id": 1,
        "cluster_id": 8,
        "command": "stop",
        "args": [
            0,
            0
        ],
        "params": {
            "options_mask": 0,
            "options_override": 0
        }
    },
    "origin": "LOCAL",
    "time_fired": "2022-04-13T08:44:22.162010+00:00",
    "context": {
        "id": "redacted",
        "parent_id": null,
        "user_id": null
    }
}
```

Seems that from those event only the stop is transfered to ControllerX? Behavior seems to have changed with a recent upgrade of HA, so that might come from there?
```
HA Version:  core-2022.4.3
OS version: 5.10.108
Supervisor:  Home Assistant OS 7.6  / supervisor-2022.04.0
AppDaemon:  4.1.0
ControllerX: Latest (seems to be v4.18.2)
```

**Comment by MagicMicky** [comment]
Should have searched better, seems related to https://github.com/xaviml/controllerx/issues/455

**Comment by xaviml** [comment]
Hi all,

As @MagicMicky , this is related to #455 which it is confirmed it is fixed on 2022.4.2, but I see now that some people still having problems with that version.

I just saw that Home Assistant released 2022.4.3 with this change: https://github.com/home-assistant/core/pull/69879, which theorically adds the fix. Can anyone try and confirm? I will still work on a better way to parse ZHA events for those controllers.

Regards,
Xavi M.

**Comment by mjmcg** [comment]
Upgrading to 2022.4.3 fixed it for me. I had added `"step_1_43_5_0_0": MediaPlayer.CLICK_VOLUME_DOWN,` to ikea.py when I was on 2022.4.2, but that didn't work until I upgraded to 2022.4.3. AppDaemon is triggering off the extra args.

2022-04-13 15:02:37.472563 INFO tradfri_dining_media_app: 🤖 Service: [1mmedia_player.volume_set[0m - entity_id: media_player.dining_room - volume_level: 0
2022-04-13 15:02:37.459447 INFO tradfri_dining_media_app: 🏃 Running `Predefined (click_volume_down)` now
2022-04-13 15:02:37.449570 INFO tradfri_dining_media_app: 🎮 Button event triggered: `step_1_43_5_0_0`

All good now. Thank you!

**Comment by xaviml** [comment]
Hi @MagicMicky,

Thank you for the hinters. Unfortunately, I do not have ZHA in my HA to test all this, so I will need to trust testers for this. I am currently changing the code so there is no need to manually change the ikea.py in the systems. However, I would like to know if the rest of the mapping for the E1744 is correct. Would you mind checking all events:

- "move_1_195_0_0"
- "move_0_195_0_0"
- "stop"
- "toggle"
- "step_0_1_0"
- "step_1_1_0"

Regards,

**Comment by xaviml** [comment]
Hi all,

I created following PR (#459) with the changes for the ZHA mapping for the following devices:
- E1810
- E1743
- E1744
- W2049

Would you mind reviewing it? I will create a beta version after revision for everyone to try before I roll new version.

Regards,
Xavi M.

**Comment by MagicMicky** [comment]
@xaviml 
Just tested the events, and this works for me:
```
            "move_1_195_0_0": MediaPlayer.HOLD_VOLUME_DOWN,
            "move_0_195_0_0": MediaPlayer.HOLD_VOLUME_UP,
            "stop": MediaPlayer.RELEASE,
            "toggle": MediaPlayer.PLAY_PAUSE,
            "step_0_1_0_0_0": MediaPlayer.NEXT_TRACK,
            "step_1_1_0_0_0": MediaPlayer.PREVIOUS_TRACK,
```

So the `step_` also benefits from added arguments.

Do you think it's safe to change this at the controllerx level? Won't there potentially be a change reverting to the previous behavior in ZHA?

**Comment by xaviml** [comment]
Hi @MagicMicky,

From what I understood from the HA issues (https://github.com/home-assistant/core/issues/69765, https://github.com/home-assistant/core/issues/69375) that the device triggering was not working. [This PR](https://github.com/home-assistant/core/pull/69879) fixed that, but the args will not go back as it was. ControllerX joins the command and the args with underscores ("_"). What I understand then is that they will not revert the args as they were, it is not the first time this happens. There is no standard event for each controller, it changes from time to time.

For now, I added this change to the PR #459, and I will release it in the next ControllerX version. If they happen to go back, I will revert the changes.

Thank you,
Xavi M.

**Comment by MagicMicky** [comment]
Perfectly makes sense, thanks a lot @xaviml for the work!

**Comment by puddly** [comment]
> ControllerX joins the command and the args with underscores ("_").

The number of "`arg`"s the remote sends unfortunately depends on its firmware version so I wouldn't rely on all of the command arguments being stable. Instead, check out the `params` dictionary in the event: you likely only need to match against `move_mode`, [which has two distinct values](https://github.com/zigpy/zigpy/blob/ab52cad124c681eb814ea4dc2e80ba789bf50e77/zigpy/zcl/clusters/general.py#L688-L690).

**Comment by xaviml** [comment]
Hi @puddly,

You are right. This is exactly what I was aiming, but using the `params` means that we need a specific parser for each device, so we can transform ZHA comands to more convenient commands like Z2M:

- brightness_move_down → Left turn
- brightness_move_up → Right turn
- brightness_stop → Stop rotation
- toggle → 1 click
- brightness_step_up → 2 click
- brightness_step_down → 3 click

(This is the example of E1744 mapping for Z2M). 

I would love to remove the generic ZHA parser (command + args joined with "_"). However, if we want to do this for ZHA, I would need someone to send me the whole ZHA event for each event, for each device. Even so, this will not prevent in the future to have changes since there are some from time to time. If someone is able to do so, please create a "new device" ticket with all the ZHA events for their device (1 ticket per device if possible).

Regards,
Xavi M.

**Comment by xaviml** [comment]
Hi all,

I just released [ControllerX v4.19.0b0](https://github.com/xaviml/controllerx/releases/tag/v4.19.0b0) with the ZHA mapping fix as indicated in #459. If no problems are raised with it, I will be releasing the non-beta version.

Regards,
Xavi M.
